crates/sui-verifier/src/id_leak_verifier.rs

```diff
@@ -215,35 +249,35 @@ fn pack(
     verifier: &mut IDLeakAnalysis,
     struct_def: &StructDefinition,
 ) -> Result<(), ExecutionError> {
-    for _ in 0..num_fields(struct_def) {
-        let value = verifier.stack.pop().unwrap();
-        if value == AbstractValue::ID {
-            return Err(verification_failure(
-                "ID is leaked into a struct.".to_string(),
-            ));
+    // When unpacking, an object whose struct type has key ability must have the first field as
+    // "id". That fields must come from one of the functions that creates a new UID.
+    let handle = verifier
+        .binary_view
+        .struct_handle_at(struct_def.struct_handle);
+    let num_fields = num_fields(struct_def);
+    verifier.stack_popn(num_fields - 1);
+    let last_value = verifier.stack.pop().unwrap();
+    if handle.abilities.has_key() {
+        if last_value != AbstractValue::Fresh {
+            let (cur_package, cur_module, cur_function) = verifier.cur_function();
+            let msg = format!(
+                "Invalid object creation in {cur_package}::{cur_module}::{cur_function}. \
+                Object created without a newly created UID. \
+                The UID must come directly from sui::{}::{}.",
+                OBJECT_NEW.1, OBJECT_NEW.2,
```

Comment:
This is true most of the time, but is it going to be confusing in the rare cases where one of the other methods for generating a fresh ID (the other fresh ID functions) is more appropriate?

Reply:
Yeah... I'm torn though since one of them is `public(friend)` and the other is just for tests. Thoughts?

Reply:
Maybe leave out the public(friend) one but mention the test only one but clarify that it's only for tests?